meetings/working-groups/collection-literals/CL-2023-04-05.md

```diff
@@ -0,0 +1,94 @@
+# Collection literals working group meeting for April 4, 2023
+
+## Agenda
+
+* Allow efficient construction of collections that are typically represented as arrays, particularly `ImmutableArray<T>`.
+
+## Discussion
+
+We discussed possible _construction method_ patterns that would allow the compiler to write directly to the underlying storage for collections such as `ImmutableArray<T>` and `List<T>`, and avoid extra allocations or copying in cases where the collection has a _known length_.
+
+There were two promising approaches that provide equivalent capability, with the difference of whether the underlying storage is allocated by the caller or by the collection.
+
+### Pass ownership of array
+One possible approach is a _construction method_ that takes an array as an argument and uses that array directly as the underlying storage of the collection, with the caller passing ownership of the array to the collection instance.
+
+For instance, given the following method for creating an `ImmutableArray<T>`:
+```csharp
+// Create ImmutableArray<T> using the array as the underlying storage.
+public static ImmutableArray<T> AsImmutableArray<T>(T[] array)
+{
+    // ...
+}
+```
+
+The compiler could translate `ImmutableArray<T> result = [x, ..e, y];` to:
+```csharp
+T[] __array = new T[e.Count + 2];
+
+int __i = 0;
+__array[__i++] = x;
+foreach (T __t in e)
+    __array[__i++] = __t;
+__array[__i++] = y;
+
+ImmutableArray<T> result = AsImmutableArray(__array);
+```
+
+### Write directly to storage span
+An alternative is a _construction method_ that creates a collection of a given size and returns the collection _and_ a span that refers to the underlying storage, and the caller populates the collection using the span.
```

Comment:
I'm curious to know more about why giving back a Span is going to result in better codegen. Is it because `.Add()` would inevitably add additional checks and indirections that can't be optimized away?

Reply:
yup.  Add has to go through the full public api.  which will also do things like update version counters inside the type.  The span it literally: here is the memory, write into it.  :)

Reply:
The example below should probably use `ImmutableArray<T>` rather than `List<T>`, since setting the capacity in the constructor and then using `.Add()` should be sufficient for `List<T>`.

Reply:
I've updated the example to use `ImmutableArray<T>` instead.